src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +697,232 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
 			return reader;
 		}
 
+		ConstPathMatcherDataPtr readSet( const Name &name ) const override
+		{
+			auto it = m_cachedSets.find( name );
+
+			if( it != m_cachedSets.end() )
+			{
+				return it->second;
+			}
+
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::PathMatcherDataPtr pathMatcherData = readLocalSet( name );
+
+			NameList childSetNames = readChildSetNames();
+
+			if( std::find( childSetNames.begin(), childSetNames.end(), name ) != childSetNames.end() )
+			{
+				NameList children;
+				childNames( children );
+
+				for( const auto &c : children )
+				{
+					ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+					ConstPathMatcherDataPtr childSets = nc->child( c, SceneInterface::NullIfMissing )->readSet( name );
+					pathMatcherData->writable().addPaths( childSets->readable(), {c} );
+				}
+			}
+
+			m_cachedSets[name] = pathMatcherData;
+			return pathMatcherData;
+		}
+
+		NameList setNames() const override
+		{
+			ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+			nc = nc->getRoot();
+			nc->convertTagsToSets();
+
+			IECore::CompoundDataBasePtr compoundDataBase = getSetCompound();
+
+			if( !compoundDataBase )
+			{
+				return NameList();
+			}
+
+			NameList setNames;
+			const auto &readable = compoundDataBase->readable();
+
+			for( auto it : readable )
+			{
+				setNames.push_back( it.first );
+			}
+
+			NameList childSetNames = readChildSetNames();
+
+			std::set<SceneInterface::Name> unique( setNames.begin(), setNames.end() );
+
+			for( const auto &childSetName : childSetNames )
+			{
+				unique.insert( childSetName );
+			}
+			return NameList( unique.begin(), unique.end() );
+		}
+
+		ReaderImplementation *getRoot()
+		{
+			if( m_parent )
+			{
+				return m_parent->getRoot();
+			}
+
+			return this;
+		}
+
+		void convertTagsToSets()
+		{
+			if( m_setsConverted )
+			{
+				return;
+			}
+
+			m_setsConverted = true;
+
+			IndexedIO::EntryIDList tags;
+			ConstIndexedIOPtr tagsIO = m_indexedIO->subdirectory( localTagsEntry, IndexedIO::NullIfMissing );
+			if( tagsIO )
+			{
+				tagsIO->entryIds( tags );
+			}
+
+			IndexedIO::EntryIDList descendentTags;
+			ConstIndexedIOPtr descendentTagsIO = m_indexedIO->subdirectory( descendentTagsEntry, IndexedIO::NullIfMissing );
+			if( descendentTagsIO )
+			{
+				descendentTagsIO->entryIds( descendentTags );
+			}
+
+			/// provided for backward compatibility...
+			if( tags.empty() && descendentTags.empty() )
+			{
+				ConstIndexedIOPtr tagsIO = m_indexedIO->subdirectory( tagsEntry, IndexedIO::NullIfMissing );
+				if( tagsIO )
+				{
+					tagsIO->entryIds( tags, IndexedIO::File );
+					tagsIO->entryIds( descendentTags );
+				}
+			}
+
+			if( tags.empty() && descendentTags.empty() )
+			{
+				return;
+			}
+
+			m_sets = new CompoundData();
```

Comment:
Here we seem to be initialising m_sets on every location as we traverse through everything, but the only place I can find us putting anything into m_sets is in the root in `appendToSet()`. Is that right or have I misunderstood?